<launch>
    <arg name="world_file"  default="$(env TURTLEBOT_GAZEBO_WORLD_FILE)"/>
    <arg name="base"      value="$(optenv TURTLEBOT_BASE kobuki)"/> <!-- create, roomba -->
    <arg name="battery"   value="$(optenv TURTLEBOT_BATTERY /proc/acpi/battery/BAT0)"/>  <!-- /proc/acpi/battery/BAT0 --> 
    <arg name="gui" default="true"/>
    <arg name="stacks"    value="$(optenv TURTLEBOT_STACKS hexagons)"/>  <!-- circles, hexagons --> 
    <arg name="3d_sensor" value="$(optenv TURTLEBOT_3D_SENSOR kinect)"/>  <!-- kinect, asus_xtion_pro --> 

    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="use_sim_time" value="true"/>
        <arg name="debug" value="false"/>
        <arg name="gui" value="$(arg gui)" />
        <!-- arg name="world_name" value="$(arg world_file)"/-->
    </include>

    <group ns="turtlebot1">
        <arg name="model_name" value="turtlebot1"/>

        <include file="$(find project1)/launch/kobuki.launch.xml">
            <arg name="base" value="$(arg base)"/>
            <arg name="stacks" value="$(arg stacks)"/>
            <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
            <arg name="model_name" value="$(arg model_name)"/>
            <arg name="x_pos" value="0"/>
            <arg name="y_pos" value="0"/>
            <arg name="z_pos" value="0"/>
        </include>

        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
            <param name="publish_frequency" type="double" value="30.0" />
            <param name="tf_prefix" type="string" value="$(arg model_name)"/>
        </node>
            
        <!-- Fake laser -->
        <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>
        <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
            args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
            <param name="scan_height" value="10"/>
            <param name="output_frame_id" value="$(arg model_name)/camera_depth_frame"/>
            <param name="range_min" value="0.45"/>
            <remap from="image" to="camera/depth/image_raw"/>
            <remap from="scan" to="scan"/>
        </node>
    </group> 

    <group ns="turtlebot2">
        <arg name="model_name" value="turtlebot2"/>

        <include file="$(find project1)/launch/kobuki.launch.xml">
            <arg name="base" value="$(arg base)"/>
            <arg name="stacks" value="$(arg stacks)"/>
            <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
            <arg name="model_name" value="$(arg model_name)"/>
            <arg name="x_pos" value="5"/>
            <arg name="y_pos" value="0"/>
            <arg name="z_pos" value="0"/>
       </include>

        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
            <param name="publish_frequency" type="double" value="30.0" />
            <param name="tf_prefix" type="string" value="$(arg model_name)"/>
        </node>

        <!-- Fake laser -->
       <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>
       <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
            args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
            <param name="scan_height" value="10"/>
            <param name="output_frame_id" value="$(arg model_name)/camera_depth_frame"/>
            <param name="range_min" value="0.45"/>
            <remap from="image" to="camera/depth/image_raw"/>
            <remap from="scan" to="scan"/>
        </node>
    </group> 

    <group ns="turtlebot3">
        <arg name="model_name" value="turtlebot3"/>

        <include file="$(find project1)/launch/kobuki.launch.xml">
            <arg name="base" value="$(arg base)"/>
            <arg name="stacks" value="$(arg stacks)"/>
            <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
            <arg name="model_name" value="$(arg model_name)"/>
            <arg name="x_pos" value="10"/>
            <arg name="y_pos" value="0"/>
            <arg name="z_pos" value="0"/>
       </include>

        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
            <param name="publish_frequency" type="double" value="30.0" />
            <param name="tf_prefix" type="string" value="$(arg model_name)"/>
        </node>

        <!-- Fake laser -->
        <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>
        <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
            args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
            <param name="scan_height" value="10"/>
            <param name="output_frame_id" value="$(arg model_name)/camera_depth_frame"/>
            <param name="range_min" value="0.45"/>
            <remap from="image" to="camera/depth/image_raw"/>
            <remap from="scan" to="scan"/>
        </node>
    </group> 

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find project1)/rviz/turtlebot1.rviz"/>
</launch>
